api/src/context.rs

```diff
@@ -797,27 +806,168 @@ impl Context {
         }
     }
 
-    pub fn estimate_gas_price<E: InternalError>(
-        &self,
-        ledger_info: &LedgerInfo,
-    ) -> Result<GasEstimation, E> {
-        let min_gas_unit_price = self.min_gas_unit_price(ledger_info)?;
-        let second_bucket = match self
+    fn next_bucket(&self, gas_unit_price: u64) -> u64 {
+        match self
             .node_config
             .mempool
             .broadcast_buckets
             .iter()
-            .enumerate()
-            .nth(1)
+            .find(|bucket| **bucket > gas_unit_price)
         {
-            Some(bucket) => *bucket.1,
-            None => min_gas_unit_price,
-        };
-        Ok(GasEstimation {
+            None => gas_unit_price,
+            Some(bucket) => *bucket,
+        }
+    }
+
+    fn default_gas_estimation(&self, min_gas_unit_price: u64) -> GasEstimation {
+        GasEstimation {
             deprioritized_gas_estimate: Some(min_gas_unit_price),
             gas_estimate: min_gas_unit_price,
-            prioritized_gas_estimate: Some(second_bucket),
-        })
+            prioritized_gas_estimate: Some(self.next_bucket(min_gas_unit_price)),
+        }
+    }
+
+    fn cached_gas_estimation(&self, current_epoch: u64) -> Option<GasEstimation> {
+        let cache = self.gas_estimation_cache.read().unwrap();
+        if let Some(epoch) = cache.last_updated_epoch {
+            if let Some(time) = cache.last_updated_time {
+                if let Some(estimation) = cache.estimation {
+                    // TODO: config for updated time
+                    if epoch == current_epoch && time.elapsed().as_millis() < 500 {
+                        return Some(estimation);
+                    }
+                }
+            }
+        }
+        None
+    }
+
+    pub fn estimate_gas_price<E: InternalError>(
+        &self,
+        ledger_info: &LedgerInfo,
+    ) -> Result<GasEstimation, E> {
+        let max_block_history = 120;
```

Comment:
nit: do we want this to be configurable, too?

Reply:
Makes sense 😄 Generally, I'll try add configs for things that I suspect will be worth tuning at runtime, so that I have some levers to pull in case things don't behave as expected, but I get that there's a trade-off. Your call 😃